src/set/int_set_par.rs

```diff
@@ -391,22 +391,108 @@ where
     }
 }
 
+fn _parallel_sum(v: &mut Vec<Integer>) {
+    use rayon::prelude::*;
+    if v.len() % 2 == 1 {
+        v.push(Integer::from(0));
+    }
+
+    while v.len() > 1 {
+        // length of list is always even
+        assert!(v.len() % 2 == 0);
+
+        // split the list in half and collapse the halves by summing
+        let split_point = v.len() / 2;
+        let (fst, snd) = v.split_at_mut(split_point);
+        fst.par_iter_mut()
+            .zip(snd)
+            .for_each(|(f, s)| f.add_assign(s as &Integer));
+
+        // cut length of list in half, possibly padding with '0'
+        if split_point != 1 && split_point % 2 == 1 {
+            v.truncate(split_point + 1);
+            v[split_point].assign(0);
+        } else {
+            v.truncate(split_point);
+        }
+    }
+
+    if v.is_empty() {
+        v.push(Integer::from(0));
+    }
+
+    assert!(v.len() == 1);
+}
+
+fn _parallel_mul(a: &mut Integer, b: &mut Integer, nproc: usize) {
+    use gmp_mpfr_sys::gmp::mpz_swap;
+    use rayon::prelude::*;
+
+    // make sure a is the larger of the two values --- gives smaller operands to muls below
+    if b.significant_bits() > a.significant_bits() {
+        unsafe {
+            let a_ptr = a.as_raw_mut();
+            let b_ptr = b.as_raw_mut();
+            mpz_swap(a_ptr, b_ptr);
+        };
```

Comment:
Would [`std::mem::swap`](https://doc.rust-lang.org/stable/std/mem/fn.swap.html) suffice for this?

Reply:
Yeah, probably.